docs/snapshotting/snapshot-support.md

```diff
@@ -0,0 +1,300 @@
+  resumed from snapshot load memory on-demand from the snapshot and copy-on-write
+  to anonymous memory.
+- Resuming from a snapshot is optimized for speed, while taking a snapshot involves
+  some extra CPU cycles for synchronously writing dirty memory pages to the memory
+  snapshot file. Taking a snapshot of a fresh microVM, on which dirty pages tracking
+  is not enabled, results in the full contents of guest memory being written to the
+  snapshot.
+- The _memory file_ and _microVM state file_ are generated by Firecracker on snapshot
+  creation, and the disk contents are flushed to their backing files.  
+- The API calls exposing the snapshotting functionality have clear **Prerequisites**
+  that describe the requirements on when/how they should be used.
+ 
+## Snapshot versioning
+
+Firecracker snapshotting implementation offers support for microVM versioning
+(`cross-version snapshots`) in the following contexts:
+- saving snapshots at older versions (being able to create a snapshot with any version
+in the `[N, N + o]` interval, while being in Firecracker version `N+o`),
+- loading snapshots from older versions (being able to load a snapshot created by any
+Firecracker version in the `[N, N + o]` interval, in a Firecracker version `N+o`).
+
+The design supports an unlimited number of versions, the value of `o` (maximum number
+of older versions that we can restore from / save a snapshot to, from the current
+version) will be defined later.
+
+## Snapshot API
+
+Firecracker exposes the following APIs for manipulating snapshots: `Pause`, `Resume`
+and `CreateSnapshot` can be called only after booting the microVM, while `LoadSnapshot`
+is allowed only before boot.
+
+### Pausing the microVM
+
+To create a snapshot, first you have to pause the running microVM and its vCPUs with
+the following API command:
+
+```bash
+curl --unix-socket /tmp/firecracker.socket -i \
+    -X PATCH "http://localhost/vm" \
+    -H "accept: application/json" \
+    -H "Content-Type: application/json" \
+    -d "{
+            "state": "Paused"
+    }"
+```
+
+**Prerequisites**: The microVM is booted.
+                   Successive calls of this request keep the microVM in the `Paused`
+                   state.
+**Effects**:
+- _on success_: microVM is guaranteed to be `Paused`.
+- _on failure_: no side-effects.
+
+## Creating snapshots
+
+Now that the microVM is paused, you can create a snapshot, which can be either a `full`
+one or a `diff` one. Full snapshots always create a complete, resume-able snapshot of
+the current microVM state and memory. Diff snapshots save the current microVM state
+and the memory dirtied since the last snapshot (full or diff). Diff snapshots are not
+resume-able, but can be merged into a full snapshot using external (provided) tooling.
+
+### Creating full snapshots
+
+For creating a full snapshot, you can use the following API command:
+
+```bash
+curl --unix-socket /tmp/firecracker.socket -i \
```

Comment:
Same here but there are also slashes that need to be removed.

Reply:
Many thanks for the heads-up! The document should be more consistent now.
I also added a note that specifies the fact that if the files indicated by the snapshot and memory paths don't exist, then they will be created.